lib/rtp_vp9/vp9_payload_descriptor.ex

```diff
@@ -287,61 +264,59 @@ defmodule Membrane.RTP.VP9.PayloadDescriptor do
             | next_dims
           ], rest}}
 
-      _ ->
+      _error ->
         {:error, :malformed_data}
     end
   end
 
-  defp ss_get_widths_heights(<<n_s::3, 1::1, _g::1, _::3>>, rest, count, dimensions)
+  defp ss_get_widths_and_heights(<<n_s::3, 1::1, _g::1, _::3>>, rest, count, dimensions)
        when count == n_s + 1,
        do: {:ok, {dimensions, rest}}
 
-  defp ss_get_widths_heights(_, _, _, _), do: {:error, :malformed_data}
+  defp ss_get_widths_and_heights(_ss_header, _rest, _count, _dimensions),
+    do: {:error, :malformed_data}
 
   defp ss_get_pg_descriptions(<<_n_s::3, _y::1, 1::1, _::3>>, rest) do
     <<n_g, rest::binary()>> = rest
 
-    1..n_g
-    |> Enum.reduce({:ok, {[], rest}}, fn _i, acc ->
-      case acc do
-        {:ok, {accumulated, rest}} ->
-          case ss_get_pg_description(rest) do
-            {:ok, {pg_description, rest}} ->
-              {:ok, {accumulated ++ [pg_description], rest}}
-
-            _ ->
-              {:error, :malformed_data}
-          end
-
-        _ ->
-          {:error, :malformed_data}
-      end
-    end)
+    with {descriptions, {:ok, rest}} <-
+           1..n_g
+           |> Enum.map_reduce({:ok, rest}, fn _i, acc ->
+             with {:ok, rest} <- acc,
+                  {:ok, {pg_description, rest}} <- ss_get_pg_description(rest) do
+               {pg_description, {:ok, rest}}
+             else
+               _error -> {nil, {:error, :malformed_data}}
+             end
+           end) do
+      {:ok, {descriptions, rest}}
+    else
+      _error -> {:error, :malformed_data}
+    end
   end
 
-  defp ss_get_pg_descriptions(_, rest), do: {:ok, {[], rest}}
+  defp ss_get_pg_descriptions(_ss_header, rest), do: {:ok, {[], rest}}
 
   defp ss_get_pg_description(<<_tid::3, _u::1, r::2, _::2, rest::binary()>>)
        when byte_size(rest) < r,
        do: {:error, :malformed_data}
 
-  defp ss_get_pg_description(<<tid::3, u::1, r::2, _::2, rest::binary()>>) do
-    pg_description = %PGDescription{tid: tid, u: u}
-
-    case r do
-      0 ->
-        {pg_description, rest}
+  defp ss_get_pg_description(<<tid::3, u::1, 0::2, _::2, rest::binary()>>),
+    do: {%PGDescription{tid: tid, u: u}, rest}
 
-      _ ->
-        <<p_diffs::binary-size(r), rest::binary()>> = rest
+  defp ss_get_pg_description(<<tid::3, u::1, r::2, _::2, rest::binary()>>)
+       when byte_size(rest) > r do
+    pg_description = %PGDescription{tid: tid, u: u}
 
-        {:ok,
-         {:binary.bin_to_list(p_diffs)
-          |> Enum.reduce(pg_description, fn p_diff, acc ->
-            %{acc | p_diffs: acc.p_diffs ++ [p_diff]}
-          end), rest}}
+    with {p_diffs, rest} <-
+           1..r
+           |> Enum.map_reduce(rest, fn _i, acc ->
+             <<p_diff, rest::binary()>> = acc
+             {p_diff, rest}
+           end) do
```

Comment:
- this will always match, so `with` is useless
- you can firstly extract `p_diffs` and `rest` with binary match and then convert `p_diffs` to list using `:binary.bin_to_list`